source/logrepl/combined.go

```diff
@@ -0,0 +1,177 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse LSN in position: %w", err)
+	}
+
+	cdcIterator, err := NewCDCIterator(ctx, &pool.Config().ConnConfig.Config, CDCConfig{
+		LSN:             lsn,
+		SlotName:        conf.SlotName,
+		PublicationName: conf.PublicationName,
+		Tables:          conf.Tables,
+		TableKeys:       conf.TableKeys,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create CDC iterator: %w", err)
+	}
+
+	c := &CombinedIterator{
+		cdcIterator:    cdcIterator,
+		activeIterator: cdcIterator,
+	}
+
+	// Resume with completed snapshot and CDC streaming
+	if conf.WithSnapshot && pos.Type == position.TypeCDC {
+		sdk.Logger(ctx).Info().Msg("Resuming in CDC mode")
+
+		if err := c.cdcIterator.Start(ctx); err != nil {
+			return nil, fmt.Errorf("failed to start CDC iterator: %w", err)
+		}
+
+		return c, nil
+	}
+
+	if conf.WithSnapshot {
+		if err := c.initSnapshotIterator(ctx, pool, conf); err != nil {
+			return nil, err
+		}
+	}
+
+	return c, nil
+}
+
+func (c *CombinedIterator) Next(ctx context.Context) (sdk.Record, error) {
+	r, err := c.activeIterator.Next(ctx)
+	if err != nil {
+		// Snapshot iterator is done, handover to CDC iterator
+		if errors.Is(err, snapshot.ErrIteratorDone) {
+			return c.useCDCIterator(ctx)
+		}
+
+		return sdk.Record{}, fmt.Errorf("failed to fetch next record: %w", err)
+	}
+
+	return r, nil
+}
+
+func (c *CombinedIterator) Ack(ctx context.Context, p sdk.Position) error {
+	return c.activeIterator.Ack(ctx, p)
+}
+
+func (c *CombinedIterator) Teardown(ctx context.Context) error {
+	logger := sdk.Logger(ctx)
+
+	var errs []error
+
+	if c.cdcIterator != nil {
+		if err := c.cdcIterator.Teardown(ctx); err != nil {
+			logger.Warn().Err(err).Msg("Failed to tear down cdc iterator")
+			errs = append(errs, fmt.Errorf("failed to teardown cdc iterator: %w", err))
+		}
+	}
+
+	if c.snapshotIterator != nil {
+		if err := c.snapshotIterator.Teardown(ctx); err != nil {
+			logger.Warn().Err(err).Msg("Failed to tear down snapshot iterator")
+			errs = append(errs, fmt.Errorf("failed to teardown snapshot iterator: %w", err))
+		}
+	}
+
+	return errors.Join(errs...)
+}
+
+func (c *CombinedIterator) initSnapshotIterator(ctx context.Context, pool *pgxpool.Pool, conf Config) error {
+	snapshotIterator, err := snapshot.NewIterator(ctx, pool, snapshot.Config{
+		Position:     conf.Position,
+		Tables:       conf.Tables,
+		TableKeys:    conf.TableKeys,
+		TXSnapshotID: c.cdcIterator.TXSnapshotID(),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create snapshot iterator: %w", err)
+	}
+
+	sdk.Logger(ctx).Info().Msg("Initial snapshot requested, starting..")
+
+	c.snapshotIterator = snapshotIterator
+	c.activeIterator = snapshotIterator
+
+	return nil
+}
+
+func (c *CombinedIterator) useCDCIterator(ctx context.Context) (sdk.Record, error) {
+	sdk.Logger(ctx).Debug().
+		Msg("snapshot done, switching to CDC mode")
+
+	c.activeIterator, c.snapshotIterator = c.cdcIterator, nil
```

Comment:
Maybe we should tear down the snapshot iterator here?

Reply:
We can, but we will need to re-create it when new tables are added, this is the safer thing to do. Will get it done.

Reply:
What @hariso said ☝️

Reply:
this was done